### DISCUSSION TITLE
Parallel Data Collection problem

### DISCUSSION THREAD

**Discussion by dxyzx0** [ds4dm/ecole#251] [Questions] [answered]
I want to reproduce the result in the paper(10000 for training, 20000 for validation) by running the example `branching-imitaion.ipynb`, but I find that the generation process is too slow to finish.
For example, this follows the setting in the paper for Capacitated Facility Location problem with `n_customer=100` and I shrink the training/validation set to `12000`, but still too slow to collect all the instances.
![image](https://user-images.githubusercontent.com/28868457/134805210-e318c89c-ca7a-4529-bd40-5a8f503be04d.png)
My cpu is `Intel(R) Xeon(R) Gold 6226R CPU @ 2.90GHz`, which I think is enough.

> Users that are interested in reproducing competitive performance should use a larger sample size, such as the n=100,000 samples used for training in the paper. In this case, we strongly recommend to parallelize data collection, as in the original Gasse et al. (2019) code.

I know I need to parallelize data collection, but how can I accomplish this process?
I tried with `multiprocessing.Pool` but I get the following error `TypeError: cannot pickle 'ecole.core.reward.IsDone' object`.

**Comment by AntoinePrv** [comment, marked as answer]
Hi @DingXiangYuanZhiXing 

Indeed `multiprocessing` requires pickle support, which is not fully supported by Ecole.
You can use the `threading`, which should not be (too) affected by the [GIL](https://realpython.com/python-gil/), given that it is released during computationally intensive C/C++ functions calls.

**Reply by dxyzx0** [reply]
I solved this problem by putting `Ecole.environment.Environment()` object into function body instead of function arguments. 

It turns out this method works and I haven't found any problem about it.
